### DISCUSSION TITLE
Advance only on correct answer

### DISCUSSION THREAD

**Discussion by jhee1610** [jspsych/jsPsych#1907] [Q&A] [answered]
Hi,

I am very new to jsPsych (and Java script as well) and I am trying to implement a task where participants are asked to copy a random three letter word into a text box before being able to continue. I am using the survey-text plugin and am having trouble with retrieving their response and then checking whether it is the same as the three letter stimulus they were displayed. This is the code i have written.

```js
 for(var i = 0; i< 3 ; i++){
            var statement = true
            var num = generateRandomStimulus()
            var acc = false
            var task = {
                type: "survey-text",
                questions: [{prompt: num}],
                preamble: "Re-Type the below characters in the box",
                on_finish: function(data) {
                    var responses = JSON.parse(data.response);
                    var answer = responses.Q0;
                    jsPsych.data.addProperties({Participant_answer: answer});

                    if (jsPsych.data.Participant_answer == num) {
                        acc = true
                        statement = false
                    } else { acc = false};
                }
                };
            
                var loop_node = {
                    timeline: [task],
                    loop_function: function(){
                        if(acc){
                            return false;
                        } else {
                            return true;
                        }
                    }
                }
            timeline.push(loop_node)
            
        }
```

I keep getting the Uncaught "SyntaxError: Unexpected token u in JSON at position 0" message when i run this. I think this is because data.response is not the correct object to pass into the JSON.parse function. I am unsure how to fix this.
Any help would be greatly appreciated.

**Comment by becky-gilbert** [comment, marked as answer]
Hi @jhee1610, sorry that validating survey-text responses is a little tricky right now. We're planning to update the survey-* plugins in the near future (later this summer or fall), and when we do that, this will be come a lot easier!

In the meantime, I had a look at your code and found a few issues. (I'm assuming you're using jsPsych v6.3.0+, if not then you may need to tweak my solution).

1. The value of `num` isn't being saved to the trial data on each iteration of your loop. The reason is that all of the loop code runs before the experiment starts, but the code _inside of a function parameter_, such as the `on_finish` function, doesn't run until during the experiment. So the `num` variable in this line: `if (jsPsych.data.Participant_answer == num) {` isn't replaced with the different values it had was when the loop was running - instead it just remains as the variable `num`. This means that when the `on_finish` function runs during the experiment, its value will always be whatever the last value of `num` was when the page loaded. (If this is confusing then it might help to read the [docs page about dynamic parameters](https://www.jspsych.org/overview/dynamic-parameters/)).
Anyway, the solution to this problem is to save the current value of `num` to the data for each trial when the trial is created. As long as we're not putting the `num` variable inside of a function, the value of `num` will be replaced with its current value during the loop. We can do this using the [trial's `data` parameter](https://www.jspsych.org/overview/plugins/#the-data-parameter):
```js
var task = {
  type: "survey-text",
  // other parameters ...
  data: {num: num}  // add the current value of 'num' to the data for this trial
}
```
2. As of jsPsych v6.3.0, objects in the jsPsych data are no longer stored as JSON-formatted strings. Instead, they remain JavaScript objects. This means that you no longer need to use `JSON.parse` to get the participant's response from the trial data. I think this is the source of the syntax error you're getting. To get the participant's response from the data in the `on_finish` function, you can do this: 
```js
var answer = data.response.Q0;
```
3. I noticed that you're storing the participant's response in the jsPsych data with this line: `jsPsych.data.addProperties({Participant_answer: answer});`. This will store the participant's response in the data for _every_ trial, not just the current trial. The variable names in the data have to be unique, so if you store multiple responses with the same variable name (Participant_answer) in the same trials, then the value of this variable will be overwritten each time. This means that all trials will contain the Participant_answer variable with the last response that was made. I'm not sure if that's what you want? If not then I would remove this line.

4. I suggest storing the response from each trial in the trial data. The `data` object that is passed into the `on_finish` function is modifiable, which means you can add new variables to it. So you can extract the response and add it to the data like this:
```js
on_finish: function(data) {
  var answer = data.response.Q0; 
  data.answer = answer; 
  // etc. ...
}
```
5. When checking the response accuracy, your current method of getting the participant's answer from the data won't work. To extract the response from the data, you would need to use something like `jsPsych.data.get()`, and then get one of the trials, get the values from that trial, and then get the value of the Participant_answer variable (see the [data overview](https://www.jspsych.org/overview/data/#storing-data-in-jspsychs-data-structure) and [data reference](https://www.jspsych.org/core_library/jspsych-data/) docs pages for more info). Since you've already extracted the participant's answer on that trial in the `on_finish` function, I think the easiest solution is to compare that value with the value of `num` from the trial data, like this:
```js
if (answer == data.num) {
```

So here's my complete suggestion for your code:
```js
for(var i = 0; i< 3 ; i++){
  var statement = true
  var num = 'abc';
  var acc = false
  var task = {
    type: "survey-text",
    questions: [{prompt: num}],
    preamble: "Re-Type the below characters in the box",
    data: {num: num}, // add the current value of num to the trial data
    on_finish: function(data) {
      var answer = data.response.Q0; // no need for JSON.parse here
      data.answer = answer; // add answer to trial data
      // only use the line below if you want to add the answer to the data to all trials - doing this will overwrite the previous answers each time
      // jsPsych.data.addProperties({Participant_answer: answer});  
      if (answer == data.num) {
        acc = true
        statement = false
      } else { 
        acc = false
      };
    }
  };

  var loop_node = {
    timeline: [task],
    loop_function: function(){
      if(acc){
        return false;
      } else {
        return true;
      }
    }
  }
  timeline.push(loop_node)
    
}
```

**Reply by jhee1610** [reply]
Thankyou so much this has helped me a lot. Thanks for taking the time to go into detail with all the elements I really appreciate it.
